components/renku_data_services/session/db.py

```diff
@@ -876,6 +876,51 @@ async def update_build(self, user: base_models.APIUser, build_id: ULID, patch: m
 
         return build_model
 
+    async def get_build_logs(
+        self, user: base_models.APIUser, build_id: ULID, max_log_lines: int | None = None
+    ) -> dict[str, str]:
+        """Get the logs of a build by querying ShipWright."""
+        if not user.is_authenticated or user.id is None:
+            raise errors.UnauthorizedError(message="You do not have the required permissions for this operation.")
+
+        async with self.session_maker() as session, session.begin():
+            stmt = select(schemas.BuildORM).where(schemas.BuildORM.id == build_id)
+            result = await session.scalars(stmt)
+            build = result.one_or_none()
+
+            if build is None:
+                raise errors.MissingResourceError(
+                    message=f"Build with id '{build_id}' does not exist or you do not have access to it."
+                )
+
+            if build.environment.environment_kind == models.EnvironmentKind.GLOBAL:
+                authorized = True
+            else:
+                launcher = await session.scalar(
+                    select(schemas.SessionLauncherORM).where(
+                        schemas.SessionLauncherORM.environment_id == build.environment_id
+                    )
+                )
+                if launcher is None:
+                    authorized = False
+                else:
+                    authorized = await self.project_authz.has_permission(
+                        user, ResourceType.project, launcher.project_id, Scope.WRITE
+                    )
+            if not authorized:
+                raise errors.MissingResourceError(
+                    message=f"Build with id '{build_id}' does not exist or you do not have access to it."
+                )
+
+        build_model = build.dump()
+
+        if self.shipwright_client is None:
+            raise errors.MissingResourceError(message=f"Build with id '{build_id}' does not have logs.")
```

Comment:
question: shouldn't this be a 5xx server error? this happening would mean the server is not set up correctly?

Reply:
For now, `shipwright_client` is meant to be `None` in two cases: when the feature is not enabled and when running tests. In the future, it would be nicer to setup mocks like in `components/renku_data_services/k8s/clients.py` to improve tests.
In https://github.com/SwissDataScienceCenter/renku-data-services/pull/656 (not ready yet), it should not be possible to run code which would touch shipwright when the feature is not enabled (the routes are not enabled).